PropertyChanged.Fody.Analyzer/SourceGeneratorEngine.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+#pragma warning disable CS0067
+static class SourceGeneratorEngine
+{
+    const string sourceFileHintName = "PropertyChanged.g.cs";
+
+    public static void GenerateSource(SourceProductionContext context, Compilation compilation, Configuration configuration, ImmutableArray<ClassDeclarationSyntax> classes)
+    {
+        var cancellationToken = context.CancellationToken;
+
+        if (configuration.IsDisabled)
+        {
+            context.AddSource(sourceFileHintName, @"// Source generator is disabled by configuration.");
+            return;
+        }
+
+        var eventInvokerName = configuration.EventInvokerName.NullIfEmpty() ?? "OnPropertyChanged";
+
+        var codeBuilder = new CodeBuilder();
+
+        try
+        {
+            codeBuilder
+                .Add("// <auto-generated/>")
+                .Add("#nullable enable")
+                .Add("#pragma warning disable CS0067")
+                .Add("using System.ComponentModel;")
+                .Add("using System.Runtime.CompilerServices;");
+
+            var classesByNamespace = classes.GroupBy(item => item.SyntaxTree)
```

Comment:
There can be several `ClassDeclarationSyntax` for a single class here if the user writes several parts manually.